### ISSUE TITLE
Reconsider major.minor pinning

### ISSUE THREAD

**Issue by blakeembrey** [Microsoft/types-publisher#12]
I brought this up before offline, but I want to post it here to discuss it thoroughly. The current implementation pins to `major.minor` which blocks people from typing anything with a smaller version. This is a bit of an issue for anyone wanting to type/use pre-release software, or if something needs to be added for an actual patch release (though that breaks semantic versioning, not everyone follows it correctly - including TypeScript - and it could occur more likely on `0.x` versions where semver doesn't actually apply).

The way this is solved in Typings is by using the build metadata part of semver. Specifically, `2.0.0+timestamp`.

**Comment by RyanCavanaugh** [comment]
I don't see how this is going to be an issue in practice (or I'm misunderstanding what you're saying here). Is there anyone who actually needs to target a specific *patch* version of a library with a specific typings file? That seems theoretical at best.

**Comment by blakeembrey** [comment]
Yes. Gulp is currently writing an alpha, and someone recently contributed typings for it.

**Comment by RyanCavanaugh** [comment]
It's an alpha, without a major version bump, with breaking changes?

**Comment by blakeembrey** [comment]
No, it'd be `4.0.0-alpha` - but I assume you don't want people installing `4.0.0-alpha` typings just because `4.0.0-beta` is now out with more breaking changes.

**Comment by blakeembrey** [comment]
Or the granularity does not matter? It's the same as though `4.0` stable is just released, you won't be able to differentiate between `4.0.0` and `4.0.0-alpha` in case the community has not caught up.

**Comment by blakeembrey** [comment]
Also, I think you'll need to do some tagging trickier here too - otherwise people doing `npm install @types/gulp` would currently be receiving the alpha typings of `4.0` (because the version has been changed to `4.0.0`) while everyone should probably be using `3.x` typings.

**Comment by RyanCavanaugh** [comment]
We can funnel through any prerelease metadata, so if the types file says it's for `4.0.0-alpha` we'll publish it at `4.0.x-alpha` and `@latest` (which would be the default) would be published for the newest version lacking a prerelease tag.

**Comment by blakeembrey** [comment]
Seems reasonable. Still a little scared of it since it's against semver and I haven't though about what might happen, but at least proxying pre-release info is a plus 👍

**Comment by blakeembrey** [comment]
Now that even TypeScript has been bitten by the bad semver implementation with https://github.com/Microsoft/TypeScript/issues/14555, is it possible to open this discussion up to finding a solution? Using the pre-release metadata might be a possible approach using NPM if there's a way to ensure breaking changes in the pre-release don't satisfy (`^`).

**Comment by blakeembrey** [comment]
Actually, I think I just realised how this can be done with NPM and zero changes required. According to https://docs.npmjs.com/misc/semver#prerelease-tags and [`semver`](https://www.npmjs.com/package/semver). Use the patch _and_ prelease versions. The pre-release represents revisions to the definition, while the patch becomes the mechanism for communicating breaking changes.

```js
semver.maxSatisfying(['7.0.2-v1', '7.0.1-v1'], '^7.0.1-v1')
```
